utopia-remix/app/handlers/validators.spec.ts

```diff
@@ -0,0 +1,160 @@
+import { prisma } from '../db.server'
+import {
+  createTestProject,
+  createTestProjectAccess,
+  createTestSession,
+  createTestUser,
+  newTestRequest,
+  truncateTables,
+} from '../test-util'
+import { validateProjectAccess } from './validators'
+import * as permissionsService from '../services/permissionsService.server'
+import { AccessLevel, UserProjectPermission } from '../types'
+import { ApiError } from '../util/errors'
+import type { ValidationResult } from '../util/api.server'
+import { Status } from '../util/statusCodes'
+
+describe('validators', () => {
+  describe('validateProjectAccess', () => {
+    afterEach(async () => {
+      await truncateTables([
+        prisma.projectID,
+        prisma.projectAccess,
+        prisma.project,
+        prisma.userDetails,
+        prisma.persistentSession,
+      ])
+
+      jest.spyOn(permissionsService, 'hasUserProjectPermission').mockRestore()
+    })
+
+    beforeEach(async () => {
+      await createTestUser(prisma, { id: 'bob' })
+      await createTestUser(prisma, { id: 'alice' })
+      await createTestSession(prisma, { key: 'bob-key', userId: 'bob' })
+      await createTestSession(prisma, { key: 'alice-key', userId: 'alice' })
+      await createTestProject(prisma, { id: 'one', ownerId: 'bob' })
+    })
+
+    const perm = UserProjectPermission.CAN_COMMENT_PROJECT // just any of the permissions is fine
+
+    afterAll(async () => {
+      jest.restoreAllMocks()
+    })
+
+    it('errors if the project id is not passed', async () => {
+      const got = await validateProjectAccess(perm, { getProjectId: () => null })(
+        newTestRequest(),
+        {},
+      )
+
+      const error = mustBeApiErrorValidator(got)
+      expect(error.status).toBe(Status.BAD_REQUEST)
+    })
+
+    it('errors if the project does not exist', async () => {
+      const got = await validateProjectAccess(perm, { getProjectId: () => 'unknown' })(
+        newTestRequest(),
+        {},
+      )
+
+      const error = mustBeApiErrorValidator(got)
+      expect(error.status).toBe(Status.NOT_FOUND)
+    })
+
+    it('does nothing if the user is the owner', async () => {
+      await createTestProjectAccess(prisma, {
+        projectId: 'one',
+        accessLevel: AccessLevel.PRIVATE,
+      })
+      const got = await validateProjectAccess(perm, { getProjectId: () => 'one' })(
+        newTestRequest({ authCookie: 'bob-key' }),
+        {},
+      )
+      expect(got.ok).toBe(true)
+    })
+
+    it('does nothing if the user has access', async () => {
+      jest.spyOn(permissionsService, 'hasUserProjectPermission').mockResolvedValue(true)
```

Comment:
instead of repeating `jest.spyOn(permissionsService, 'hasUserProjectPermission').<method>` we can save the instance at the top:
`const hasUserProjectPermissionMock= jest.spyOn(permissionsService, 'hasUserProjectPermission')` 
and just use `hasUserProjectPermissionMock.<method>` everywhere.
(safer from typos, easier to read, etc)